### PR TITLE
geo/geomfn: Implements ST_ShortestLine

### LINKED ISSUE #49035: geo/geomfn: implement ST_ShortestLine({geometry,geometry})
Implement `ST_ShortestLine` on arguments {geometry,geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_ShortestLine.html).

_Observers: Please react to this issue if you need this functionality._

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end.
* Ensure the documentation is regenerated by calling `make buildshort`.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

The following additional guidance has been issued on implementing this function:

Re-use the code from #50018 for minimum distance.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-06-11T05:31:15Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by abhishek20123g**
Fixes https://github.com/cockroachdb/cockroach/issues/49035

This PR implements ST_ShortestLine({geometry, geometry}) builtin
function. Which allow us to find the LineString corresponds to
the minimum distance between given two geometrical objects.

Release note (sql change): This PR implements ST_ShortestLine({geometry,
geometry}) builtin function.

### CODE REVIEW

File: pkg/geo/geogfn/distance.go [hunk 1]
```diff
@@ -159,8 +159,10 @@ type s2GeodistEdgeCrosser struct {
 var _ geodist.EdgeCrosser = (*s2GeodistEdgeCrosser)(nil)
 
 // ChainCrossing implements geodist.EdgeCrosser.
-func (c *s2GeodistEdgeCrosser) ChainCrossing(p geodist.Point) bool {
-	return c.EdgeCrosser.ChainCrossingSign(p.(*s2GeodistPoint).Point) != s2.DoNotCross
+func (c *s2GeodistEdgeCrosser) ChainCrossing(p geodist.Point) (bool, geodist.Point) {
+	// Returns nil for the intersection point as we don't require the intersection
+	// point for calculation of minimum distance in geography.
```
Comment (otan) [inline pkg/geo/geogfn/distance.go:164]: nit: replace `for calculation of minimum distance in geography.` with `as we do not have to implement ShortestLine in geography.`

**Review by otan — COMMENTED** [review]
Looks neat! one comment nit

**Comment by abhishek20123g** [comment]
Small Overview
* In case of `geometry` intersecting we have to determine the intersection point and for that, I have changed `OnIntersect()` to `Onintersect(Point)`

* For intersection case between the edge also we have to determine the intersection point as same I have changed `ChainCrossing(Point) bool` to `ChainCrossing(Point) (bool, Point)` 

* There are cases in which we are getting -0 `LineString(-0 -0, -0 -0)`. I am not sure it's issuable or not. That is we have to get rid of this.

The reason we are getting -0:
```golang
a := 0.000
b := -1.000

fmt.Println(a/b)
```
will be -0
thought in golang `-0 == 0` is true but in `PostGIS`
```
gis=# select st_equals('LINESTRING(0 0, 0 0)'::geometry, 'LINESTRING(-0 -0, -0 -0)'::geometry);
 st_equals 
-----------
 f
(1 row)

```

File: pkg/geo/geogfn/distance.go (hunk 1 above)
Comment (abhishek20123g) [reply pkg/geo/geogfn/distance.go:164]: sure

**Comment by otan** [comment]
> There are cases in which we are getting -0 LineString(-0 -0, -0 -0). I am not sure it's issuable or not. That is we have to get rid of this.

Do you know when we have a negative number as a divisor?

**Comment by otan** [comment]
```
select st_equals('LINESTRING(0 0, 0 0)'::geometry, 'LINESTRING(-0 -0, -0 -0)'::geometry);
```

this isn't equal because you can't have a linestring where both points are the same, and since that is "not valid" postgis can do whatever it wants

**Review by otan — APPROVED** [review]
the -0 thing seems fine to me.

**Comment by abhishek20123g** [comment]
the negative denominator was not in `crdb`, it was present in `github.com/twpayne/go-geom/xy/lineintersector` 

https://github.com/twpayne/go-geom/blob/62c03a64717d682dd89319176fd7247944f0c518/xy/lineintersector/nonrobust_line_intersector.go#L131-L134

Thank you for the review

**Comment by abhishek20123g** [comment]
> this isn't equal because you can't have a linestring where both points are the same, and since that is "not valid" postgis can do whatever it wants

Got it!

**Comment by abhishek20123g** [comment]
Rebasing PR
seems like build fails is not from this PR.

**Pull request merged**